### PR TITLE
Version CheckpointContents

### PR DESCRIPTION
**Pull request by mystenmark**
## Description 

Stacked on #8972 

Version CheckpointContents.

Note that I didn't use the usual `#[enum_dispatch]` approach in this case because I think its unlikely we are going to extend this format any time soon. We can always add that in later if we need it.

## Test Plan 

```cargo simtest```

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-types/src/messages_checkpoint.rs [hunk 1]
```diff
@@ -327,120 +209,37 @@ impl SignedCheckpointSummary {
 // or other authenticated data structures to support light
 // clients and more efficient sync protocols.
 
+pub type CheckpointSummaryEnvelope<S> = Envelope<CheckpointSummary, S>;
 pub type CertifiedCheckpointSummary = CheckpointSummaryEnvelope<AuthorityStrongQuorumSignInfo>;
+pub type SignedCheckpointSummary = CheckpointSummaryEnvelope<AuthoritySignInfo>;
 
-impl CertifiedCheckpointSummary {
-    /// Aggregate many checkpoint signatures to form a checkpoint certificate.
-    pub fn aggregate(
```
Comment (aschran) [inline crates/sui-types/src/messages_checkpoint.rs:334]: don't entirely follow all the changes to the signature logic, maybe a second pair of eyes on that would be worthwhile (or could walk me through it on huddle?)

**Review by aschran — APPROVED** [review]

File: crates/sui-types/src/messages_checkpoint.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-types/src/messages_checkpoint.rs:334]: sorry this is from a stacked PR, let me rebase

**Review by aschran — APPROVED** [review]

File: crates/sui-types/src/messages_checkpoint.rs [hunk 2]
```diff
@@ -330,15 +357,17 @@ impl CheckpointContents {
...
-        self.transactions.len()
+        self.as_v1().transactions.len()
     }
 
-    pub fn digest(&self) -> CheckpointContentsDigest {
-        CheckpointContentsDigest::new(sha3_hash(self))
+    pub fn digest(&self) -> &CheckpointContentsDigest {
+        self.as_v1()
+            .digest
+            .get_or_init(|| CheckpointContentsDigest::new(sha3_hash(self)))
```
Comment (bmwill) [inline crates/sui-types/src/messages_checkpoint.rs:370]: How are you distinguishing between V1 and some future V2? if you're calculating the digest without including the variant byte used to distinguish between two versions of the type then i don't think we'll actually have an easy path forward for evolving this, unless this is being handled some other way.

Comment (mystenmark) [reply crates/sui-types/src/messages_checkpoint.rs:370]: oh good catch, need to compute the hash on the outer structure

Comment (mystenmark) [reply crates/sui-types/src/messages_checkpoint.rs:370]: wait, no, this i fine - the hash is over `self`, even though it is stored in V1, so it does include the variant tag.

Comment (bmwill) [reply crates/sui-types/src/messages_checkpoint.rs:370]: Oh...the as_v1 threw me off...

**Pull request merged**